ui/redux/actions/user.js

```diff
@@ -731,82 +739,65 @@ export function doUserSetReferrerWithUri(uri) {
           }
-          const isRewardApproved = selectIsRewardApproved(state);
-          dispatch(doUserSetReferrer(referrerCode, isRewardApproved));
+          dispatch(doUserSetReferrer(referrerCode));
         }
       } catch (error) {
         dispatch({
-          type: ACTIONS.USER_SET_REFERRER_FAILURE,
+          type: ACTIONS.USER_SET_REFERRER_FAIL,
           data: { error },
         });
       }
     } else {
       referrerCode = claim.permanent_url.replace('lbry://', '');
-      const isRewardApproved = selectIsRewardApproved(state);
-      dispatch(doUserSetReferrer(referrerCode, isRewardApproved));
+      dispatch(doUserSetReferrer(referrerCode));
     }
   };
 }
 
-export function doUserSetReferrer(referrer, shouldClaim) {
-  return async (dispatch, getState) => {
-    dispatch({
-      type: ACTIONS.USER_SET_REFERRER_STARTED,
-    });
+export const doUserSetReferrer = (referrerUri) => async (dispatch, getState) => {
+  let state = getState();
+  const hasSetReferrerPending = selectSetReferrerPending(state);
+  const hasReferrer = Boolean(selectReferrer(state));
 
-    let claim, referrerCode;
-    const isValid = isURIValid(referrer);
-
-    if (isValid) {
-      const state = getState();
-      const uri = normalizeURI(referrer);
-      claim = selectClaimForUri(state, uri);
-
-      if (!claim) {
-        try {
-          const response = await Lbry.resolve({ urls: [uri] });
-          if (response && response[uri] && !response[uri].error) claim = response && response[uri];
-          if (claim) {
-            if (claim.signing_channel) {
-              referrerCode = claim.signing_channel.permanent_url.replace('lbry://', '');
-            } else {
-              referrerCode = claim.permanent_url.replace('lbry://', '');
-            }
-          }
-        } catch (error) {
-          dispatch({
-            type: ACTIONS.USER_SET_REFERRER_FAILURE,
-            data: { error },
-          });
-        }
-      } else {
-        referrerCode = claim.permanent_url.replace('lbry://', '');
-      }
-    }
+  if (hasSetReferrerPending || hasReferrer) return;
+
+  dispatch({ type: ACTIONS.USER_SET_REFERRER_START });
 
-    if (!referrerCode) {
-      referrerCode = referrer;
+  let claim;
+  let referrerCode = referrerUri;
+
+  const isValid = isURIValid(referrerUri);
+
+  if (isValid) {
+    const uri = normalizeURI(referrerUri);
+    claim = selectClaimForUri(state, uri);
+
+    if (!claim) {
+      await dispatch(doResolveUri(uri))
+        .then(() => {
+          state = getState();
+          claim = selectClaimForUri(state, uri);
+        })
+        .catch((error) => dispatch({ type: ACTIONS.USER_SET_REFERRER_FAIL, data: { error } }));
     }
 
-    try {
-      await Lbryio.call('user', 'referral', { referrer: referrerCode }, 'post');
-      dispatch({
-        type: ACTIONS.USER_SET_REFERRER_SUCCESS,
-      });
-      if (shouldClaim) {
+    referrerCode = getChannelFromClaim(claim).permanent_url.replace('lbry://', '');
```

Comment:
@saltrafael  This is failing in some cases

https://sentry.lbry.tech/organizations/odysee/issues/30368/?project=2